src/install/install.ts

```diff
@@ -0,0 +1,137 @@
+  ]);
+}
+
+function _install(bowerJson?: JsonObject, directory?: string): Promise<void> {
+  return new Promise(async (resolve, reject) => {
+    const config = defaultBowerConfig({
+      save: false,
+      directory,
+    });
+    const cwd: string = config.cwd || process.cwd();
+    const renderer = new StandardRenderer('install', {
+      cwd,
+      color: true,
+    });
+    const bowerLogger = new BowerLogger();
+
+    const project = new BowerProject(config, bowerLogger);
+
+    bowerLogger.on('log', (log: bower.LogData) => {
+      renderer.log(log);
+    });
+    bowerLogger.on('end', (data: any) => {
+      renderer.end(data);
+      resolve();
+    });
+    bowerLogger.on('error', (err: Error) => {
+      renderer.error(err);
+      reject(err);
+    });
+
+    // This is the only way I could find to provide a JSON object to the
+    // Project. It's a hack, and might break in the future, but it works.
+    if (bowerJson) {
+      (project as any)._json = bowerJson;
+      (project as any)._jsonFile = path.join(cwd, 'bower.json');
+    }
+
+    project.install([], {save: false}, config);
+  });
+}
+
+async function installDefault(): Promise<void> {
+  logger.info(`Installing Bower components`);
+  await _install();
+  logger.info(`Finished installing Bower components`);
+}
+
+async function installVariants(): Promise<void> {
+  const json = await new Promise<any>((resolve, reject) => {
+    const config = defaultBowerConfig({
+      save: false,
+    });
+    const cwd = config.cwd || process.cwd();
+    readBowerJson(cwd, {}, (err: any, json: any) => {
+      err ? reject(err) : resolve(json);
+    });
+  });
+  const variants = json['variants'];
+  if (variants) {
+    await Promise.all(Array.from(Object.keys(variants)).map(async (variantName) => {
+      const variant = variants[variantName];
+      const variantBowerJson = mergeJson(variant, json);
+      const variantDirectory = `bower_components-${variantName}`;
+      logger.info(`Installing variant ${variantName} to ${variantDirectory}`);
+      await _install(variantBowerJson, variantDirectory);
+      logger.info(`Finished installing variant ${variantName}`);
+    }));
+  }
+}
+
+function mergeJson(from: JsonObject, to: JsonObject): JsonObject {
+  // First, make a shallow copy of `to` target
+  const merged = Object.assign({}, to);
+
+  // Next, merge in properties from `from`
+  for (const key in from) {
+    // TODO(justinfagnani): If needed, we can add modifiers to the key
+    // names in `from` to control merging:
+    //   * "key=" would always overwrite, not merge, the property
+    //   * "key|" could force a union (merge), even for Arrays
+    //   * "key&" could perform an intersection
+    const fromValue = from[key];
+    const toValue = to[key];
+    if (isPrimitiveOrArray(fromValue) || isPrimitiveOrArray(toValue)) {
+      merged[key] = fromValue;
+    } else {
+      merged[key] = mergeJson(fromValue as JsonObject, toValue as JsonObject);
```

Comment:
yea, but JavaScript... and user-provided data... ☠️ 

I only bring it up because I've been bit by this in the past when at the time it was impossible. If it's not hard to do, I'd still prefer to see it organized that way.